mlfinlab/data_structures/imbalance_data_structures.py

```diff
@@ -0,0 +1,287 @@
+
+        if tick_diff > 0:
+            tick_rule = 1.0
+        elif tick_diff < 0:
+            tick_rule = -1.0
+        else:
+            tick_rule = prev_tick_rule
+
+        dollar_imb = tick_rule * volume * price
+        volume_imb = tick_rule * volume
+        # cumulative dollar imbalance up till now (reset on every imbalance bar)
+
+        cum_tick_imb += tick_rule
+        cum_dollar_imb += dollar_imb
+        cum_volume_imb += volume_imb
+
+        if metric == 'tick_imbalance':
+            imbalance_array.append(tick_rule)  # latest relevant imbalances
+            theta_imb = cum_tick_imb
+        elif metric == 'dollar_imbalance':
+            imbalance_array.append(dollar_imb)
+            theta_imb = cum_dollar_imb
+        elif metric == 'volume_imbalance':
+            imbalance_array.append(volume_imb)
+            theta_imb = cum_volume_imb
+
+        imb_flag = False
+        if len(imbalance_array) < exp_num_ticks:
+            exp_tick_imb = np.nan  # waiting for array to fill for ewma
+        else:
+            # expected imbalance per tick
+            ewma_window = int(exp_num_ticks * num_prev_bars)
+            exp_tick_imb = ewma(
+                np.array(imbalance_array[-ewma_window:], dtype=float), window=ewma_window)[-1]
+            if np.abs(theta_imb) > exp_num_ticks * np.abs(exp_tick_imb):
+                imb_flag = True
+
+        # Check min max
+        if price > high_price:
+            high_price = price
+        if price <= low_price:
+            low_price = price
+
+        prev_price = price  # update prev_price
+        prev_tick_rule = tick_rule
+
+        #print(date_time, cum_dollar_imb, exp_num_ticks, exp_tick_imb)
+        # If threshold reached then take a sample
+        if imb_flag is True:   # pylint: disable=eval-used
+            # Create bars
+            open_price = cache[0][1]
+            low_price = min(low_price, open_price)
+            close_price = price
+            num_ticks_bar.append(cum_ticks)
+            expected_num_ticks_bar = ewma(
+                np.array(num_ticks_bar[-num_ticks_ewma_window:], dtype=float), num_ticks_ewma_window)[-1]  # expected number of ticks based on formed bars
+            # Update bars & Reset counters
+            list_bars.append([date_time, open_price, high_price, low_price, close_price,
+                              cum_volume, cum_dollar_value, cum_ticks])
+            cum_ticks, cum_dollar_value, cum_volume, cum_dollar_imb, cum_tick_imb, cum_volume_imb = 0, 0, 0, 0, 0, 0
+            high_price, low_price = -np.inf, np.inf
+            exp_num_ticks = expected_num_ticks_bar
+
+        # Update cache
+        cache.append([date_time, price, low_price, high_price,
+                      cum_volume, cum_dollar_value, cum_ticks, cum_dollar_imb, cum_tick_imb, cum_volume_imb, exp_num_ticks, imbalance_array])
+    return list_bars, cache, num_ticks_bar, prev_price, prev_tick_rule
+
+
+def _assert_dataframe(test_batch):
```

Comment:
The same about naming as above. Straight use is not explained. What are you asserting? It's unclear and you should go back to function definition and look for the comments. This is not effective for developers from aside. Also, you use try/except statement besides the asserts.